### PR TITLE
Span Inferrer: Improve resource name and forward errors

### PR DESCRIPTION
**Pull request by hpetru**
### What does this PR do?

- Improves the Endpoint's names on Datadog APM service page
- Mark errors for inferred spans where the lambda response code is 5XX

Endpoints before those changes:
Screen Shot 2022-03-01 at 11 29 34 AM

Endpoints after those changes:
Screen Shot 2022-03-01 at 11 30 34 AM

-----

Failed requests before those changes:

Screen Shot 2022-03-01 at 11 32 09 AM

Failed requests after those changes:
Screen Shot 2022-03-01 at 11 33 24 AM

### Motivation

Endpoints names were not shown correctly, not like other DD APM integration for other frameworks, now the Endpoint name contains the method and also the path, that makes the DD APM service page simpler and more descriptive.

Requests where lambda failed where reported as `Ok` on API Gateway service page, that's a bug.

### Types of Changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

**Comment by hpetru** [comment]
@astuyve Please take a look at those changes and let me know if any additional changes are needed to get this merged.

File: src/trace/listener.ts [hunk 1]
```diff
@@ -131,13 +131,18 @@ export class TraceListener {
...
       // Store the status tag in the listener to send to Xray on invocation completion
       this.triggerTags["http.status_code"] = statusCode!;
       if (this.tracerWrapper.currentSpan) {
         this.tracerWrapper.currentSpan.setTag("http.status_code", statusCode);
       }
       if (this.inferredSpan) {
         this.inferredSpan.setTag("http.status_code", statusCode);
+
+        if(statusCode) {
+          this.inferredSpan.setTag("error", Number(statusCode) >= 500);
```
Comment (hpetru) [inline src/trace/listener.ts:144]: @astuyve Is there a better way to assign the error from `currentSpan` to `inferredSpan` ?

Here is how the traces looks without those changes:
Screen Shot 2022-03-01 at 11 44 35 AM

As you can see the error information is not forwarded to the API Gateway inferred span

File: event_samples/api-gateway-v1.json [hunk 2]
```diff
@@ -70,7 +70,7 @@
     "requestTime": "04/Mar/2020:19:15:17 +0000",
     "requestTimeEpoch": 1583349317135,
     "resourceId": null,
-    "resourcePath": "/my/path",
+    "resourcePath": "/path",
```
[the remark this answers is not in the document]
Comment (hpetru) [reply event_samples/api-gateway-v1.json:73]: Sure, here is an example:




<details>
<summary>API Gateway V1 event payload</summary>

```json
{
   "resource":"/hello/{id}",
   "path":"/test/hello/123",
   "httpMethod":"GET",
   "headers":{
      "Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "Accept-Encoding":"gzip, deflate, br",
      "Accept-Language":"en-US,en;q=0.9",
      "CloudFront-Forwarded-Proto":"https",
      "CloudFront-Is-Desktop-Viewer":"true",
      "CloudFront-Is-Mobile-Viewer":"false",
      "CloudFront-Is-SmartTV-Viewer":"false",
      "CloudFront-Is-Tablet-Viewer":"false",
      "CloudFront-Viewer-Country":"MD",
      "Cookie":"__hssrc=1; __hstc=143039082.651a70405bbcb6e3a931dac01ee02bb4.1637577173256.1645530631053.1645726702628.22; _fbp=fb.1.1644925243108.1187820791; _ga=GA1.2.785699867.1637577172; _gid=GA1.2.1031203521.1646043414; hubspotutk=651a70405bbcb6e3a931dac01ee02bb4; messagesUtk=ed940839f19d4a0e8a1c146fd40c2bf7",
      "Host":"api.admin.slateteams.com",
      "sec-ch-ua":"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"98\", \"Google Chrome\";v=\"98\"",
      "sec-ch-ua-mobile":"?0",
      "sec-ch-ua-platform":"\"macOS\"",
      "sec-fetch-dest":"document",
      "sec-fetch-mode":"navigate",
      "sec-fetch-site":"none",
      "sec-fetch-user":"?1",
      "upgrade-insecure-requests":"1",
      "User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36",
      "Via":"2.0 28e5a05285f99582aabd539b43e9b668.cloudfront.net (CloudFront)",
      "X-Amz-Cf-Id":"FV0XD3pdQnP08n-YsQBHyzxIpseT5pYL1Iy9b5oqFlvPIU0Oxa23DQ==",
      "X-Amzn-Trace-Id":"Root=1-621e780d-420b729234dce3ba442652ad",
      "X-Forwarded-For":"89.149.78.25, 130.176.33.87",
      "X-Forwarded-Port":"443",
      "X-Forwarded-Proto":"https"
   },
   "multiValueHeaders":{
      "Accept":[
         "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
      ],
      "Accept-Encoding":[
         "gzip, deflate, br"
      ],
      "Accept-Language":[
         "en-US,en;q=0.9"
      ],
      "CloudFront-Forwarded-Proto":[
         "https"
      ],
      "CloudFront-Is-Desktop-Viewer":[
         "true"
      ],
      "CloudFront-Is-Mobile-Viewer":[
         "false"
      ],
      "CloudFront-Is-SmartTV-Viewer":[
         "false"
      ],
      "CloudFront-Is-Tablet-Viewer":[
         "false"
      ],
      "CloudFront-Viewer-Country":[
         "MD"
      ],
      "Cookie":[
         "__hssrc=1; __hstc=143039082.651a70405bbcb6e3a931dac01ee02bb4.1637577173256.1645530631053.1645726702628.22; _fbp=fb.1.1644925243108.1187820791; _ga=GA1.2.785699867.1637577172; _gid=GA1.2.1031203521.1646043414; hubspotutk=651a70405bbcb6e3a931dac01ee02bb4; messagesUtk=ed940839f19d4a0e8a1c146fd40c2bf7"
      ],
      "Host":[
         "api.admin.slateteams.com"
      ],
      "sec-ch-ua":[
         "\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"98\", \"Google Chrome\";v=\"98\""
      ],
      "sec-ch-ua-mobile":[
         "?0"
      ],
      "sec-ch-ua-platform":[
         "\"macOS\""
      ],
      "sec-fetch-dest":[
         "document"
      ],
      "sec-fetch-mode":[
         "navigate"
      ],
      "sec-fetch-site":[
         "none"
      ],
      "sec-fetch-user":[
         "?1"
      ],
      "upgrade-insecure-requests":[
         "1"
      ],
      "User-Agent":[
         "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36"
      ],
      "Via":[
         "2.0 28e5a05285f99582aabd539b43e9b668.cloudfront.net (CloudFront)"
      ],
      "X-Amz-Cf-Id":[
         "FV0XD3pdQnP08n-YsQBHyzxIpseT5pYL1Iy9b5oqFlvPIU0Oxa23DQ=="
      ],
      "X-Amzn-Trace-Id":[
         "Root=1-621e780d-420b729234dce3ba442652ad"
      ],
      "X-Forwarded-For":[
         "89.149.78.25, 130.176.33.87"
      ],
      "X-Forwarded-Port":[
         "443"
      ],
      "X-Forwarded-Proto":[
         "https"
      ]
   },
   "queryStringParameters":null,
   "multiValueQueryStringParameters":null,
   "pathParameters":{
      "id":"123"
   },
   "stageVariables":null,
   "requestContext":{
      "resourceId":"3wyzls",
      "resourcePath":"/hello/{id}",
      "httpMethod":"GET",
      "extendedRequestId":"OUeyJF0OIAMFdrA=",
      "requestTime":"01/Mar/2022:19:46:21 +0000",
      "path":"/test/hello/123",
      "accountId":"311994741943",
      "protocol":"HTTP/1.1",
      "stage":"dev",
      "domainPrefix":"api",
      "requestTimeEpoch":1646163981519,
      "requestId":"16604eac-e7a0-415c-bdc3-2daae035db00",
      "identity":{
         "cognitoIdentityPoolId":null,
         "accountId":null,
         "cognitoIdentityId":null,
         "caller":null,
         "sourceIp":"89.149.78.25",
         "principalOrgId":null,
         "accessKey":null,
         "cognitoAuthenticationType":null,
         "cognitoAuthenticationProvider":null,
         "userArn":null,
         "userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36",
         "user":null
      },
      "domainName":"api.admin.slateteams.com",
      "apiId":"a5djw34ew4"
   },
   "body":null,
   "isBase64Encoded":false
}
```

</details>

<details>
<summary>serverless.yml</summary>

```yml
service: api-demo

frameworkVersion: '2'

plugins:
  - serverless-plugin-datadog

provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221

functions:
  hello:
    handler: handler.hello
    events:
        - http: 
            path: hello/{id}
            method: get
            request: 
              parameters: 
                paths: 
                  id: true

  goodbye:
    handler: handler.goodbye
    events:
        - http: GET /goodbye

custom:
  datadog:
    addLayers: true
    apiKey: *******
    flushMetricsToLogs: false
    enableXrayTracing: true

```
</details>

File: src/trace/listener.ts (hunk 1 above)
Comment (hpetru) [reply src/trace/listener.ts:144]: Got it, can you please help me understand how I can actually get the error from the `wrappedCurrentSpan` ? I don't think that's a really good idea to actually set the error tag based on the lambda status code.

File: src/trace/span-inferrer.ts [hunk 3]
```diff
@@ -56,13 +56,15 @@ export class SpanInferrer {
...
+    const resourcePath = event.rawPath || event.requestContext.resourcePath || event.requestContext.routeKey;
+
     let method;
     if (event.requestContext.httpMethod) {
       method = event.requestContext.httpMethod;
     } else if (event.requestContext.http) {
       method = event.requestContext.http.method;
     }
-    const resourceName = [domain, path].join(" ");
+    const resourceName = [method || domain, resourcePath].join(" ");
```
[the remark this answers is not in the document]
Comment (hpetru) [reply src/trace/span-inferrer.ts:67]: Great, leaving this one to you, but IMO those changes cleans up a lot the Endpoints list

**Comment by hpetru** [comment]
@astuyve After some additional testing I discovered that my changed from this commit https://github.com/DataDog/datadog-lambda-js/pull/267/commits/af213b46970d253b9146c63ab71c491da12127c7 doen't work as expected, failed requests still doen't have the `error` tag

Screen Shot 2022-03-02 at 12 22 39 PM

**Comment by hpetru** [comment]
> @astuyve After some additional testing I discovered that my changed from this commit https://github.com/DataDog/datadog-lambda-js/commit/af213b46970d253b9146c63ab71c491da12127c7 doen't work as expected, failed requests still doen't have the error tag

Never-mind, I figured it out, now error tag is set before the inferred span is finished, ref: https://github.com/DataDog/datadog-lambda-js/pull/267/commits/dab71d1aeb79a8b1eb04c40c72f4f8795bb89333

**Comment by hpetru** [comment]
@astuyve Looking forward for your feedback, I really want to get all those changed released, since is our main blocker for fully adoption DataDog

File: src/trace/span-inferrer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (astuyve) [reply src/trace/span-inferrer.ts:67]: The only reason I can come up with for this was that I misunderstood feedback from early testers. I think this is good to go

**Comment by astuyve** [comment]
Hi! Yes, thanks @hpetru, sorry I was working on this https://github.com/DataDog/serverless-plugin-datadog/issues/232

I think this is good to go, will merge and release ASAP.

File: event_samples/api-gateway-v1.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (astuyve) [reply event_samples/api-gateway-v1.json:73]: I see, so right now your API Service routes are broken down by each individual resource, yikes!

**Review by astuyve — APPROVED** [review]

**Comment by astuyve** [comment]
Hi @hpetru, thanks again for your contribution. This is released in version 5.73.0, and layer version 73! I'll be updating the datadog plugin to automatically pick this change up as well.

**Pull request merged**